ユーザーフォロー課題

ham-cap opened the pull request:
ユーザーフォローの課題を提出します。
ご確認のほどよろしくお願いいたします。

ham-cap committed:
Userモデルにfollow・unfollow・following?メソッドを追加

ham-cap committed:
ログインユーザーがフォローしているユーザーの一覧表示機能を追加

ham-cap committed:
フォローボタンを設置・createアクションを実装

ham-cap committed:
フォロー関連ボタンの表示切り替えを実装・destroyアクションを実装

・ログインユーザー自身のユーザーページにはフォロー及びフォロー解除ボタンを表示しないようにした
・対象ユーザーをフォロー済みか否かによってフォロー及びフォロー解除ボタンのいずれか一方のみが表示されるようにした
・destroyアクションを実装し、フォロー解除をできるようにした

ham-cap committed:
destroyアクションを修正・日本語化

・destroyアクション内で取得できるparams[:id]の中身をfollowsテーブルのレコードのidではなくusersのidと誤認しており、誤った処理になっていたため修正した
・フォロー機能の実装にあたってviewに追加した項目を日本語化した

ham-cap committed:
ERB Lintの導入・userモデルの関連付けにinverse_ofを追加・その他rubocop, erblintの指摘事項を修正

・課題提出前のerblint実行が必須化されたため、ERB Lintを導入した。
・rubocopからuserモデル内の一部の関連付けにinverse_ofを指定するよう指摘があった。
　確認したところ、:throughや:foreign_keyが使われていると双方向の関連付けが自動で認識されなくなり、意図した挙動にならない場合があるため、inverse_ofを使用することでRailsに対して双方向の関連付けであることを明示する必要があるとのことだった。
・その他、rubocop及びerblintによる軽微な修正を自動反映させた。

maedana commented on app/assets/stylesheets/follows.scss:
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Follows controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
自動生成かなと思います。特に使ってない様子ですし削除しましょう。他にもテストコード関連で無意味なファイルが複数ありそうです。それぞれ確認して削除しましょう。

こういう話です。
https://bootcamp.fjord.jp/pages/info-for-github#meaningless-file

maedana commented on db/schema.rb:
```diff
@@ -49,6 +49,16 @@
 
+  create_table "follows", force: :cascade do |t|
+    t.integer "follower_id"
+    t.integer "followed_id"
```
上記2つの外部キーはいずれも必ず入るかなと思います。よって`not null`制約つけましょう

maedana commented on db/migrate/20220716063208_create_follows.rb:
```diff
@@ -0,0 +1,13 @@
+  def change
+    create_table :follows do |t|
+      t.integer :follower_id
+      t.integer :followed_id
```
外部キーを追加するときは原則はreferences使うほうがいいですね。

参考
https://railsguides.jp/association_basics.html#belongs-to関連付けに対応する外部キーを作成する
https://api.rubyonrails.org/v6.1.6.1/classes/ActiveRecord/ConnectionAdapters/Table.html#method-i-references
使えるオプションはこれですかね。以下にある通りデフォルトでindexも自動で作ってくれます。
https://api.rubyonrails.org/v6.1.6.1/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-add_reference

maedana commented on config/routes.rb:
```diff
@@ -3,5 +3,12 @@
+  end
+  scope ':id' do
+    resources :follows, only: %i(create destroy)
+  end
```
これはちょっと見かけないですし、トリッキーというかわかりにくいですね。
```
follows_path POST /:id/follows(.:format) follows#create
follow_path DELETE /:id/follows/:id(.:format) follows#destroy
```
というURIが生成されますが、前者は何のidかわからないですし、後者はidが2つ重複しちゃってますよね。ようするに
- 指定ユーザをログインユーザがフォローする
- 指定ユーザをログインユーザがフォロー解除する

をやりたいわけですよね。scopeなしだと
```
follows_path POST /follows(.:format) follows#create
follow_path DELETE /follows/:id(.:format) follows#destroy
```

ですが、前者はviewで`<%= hidden_field_tag :followed_id, @user.id %>`のようにしてURIではなくsubmit時にbodyとして送信してますし、後者はscopeなくてもURIに含まれるidからフォロー済みのFollowは特定出来るわけです。

以上踏まえてscopeなくした上でcontrollerの処理など見直してみて欲しいです。

maedana commented on app/views/users/show.html.erb:
```diff
@@ -29,11 +29,33 @@
 <% end %>
+  <% if current_user == @user %>
+    <%= link_to t('views.common.edit'), edit_user_registration_path %> |
+  <% end %>
```
`current_user == @user`をif/unlessでそれぞれ書いてますけど、if/elseで書けそうですね

maedana commented on app/views/users/show.html.erb:
```diff
@@ -29,11 +29,33 @@
+</p>
+<% unless current_user == @user %>
+  <% if current_user.followings.exists?(id: @user.id) %>
+    <%= form_with(model: current_user.active_relationships.find_by(followed_id: @user.id),
```
なるべくviewではSQL発行したくないですねー。この2行でフォローデータの存在チェックをしてからそれを改めて読み出してますが、1SQLで済ませたいですね。結局やってること同じですし。その上でcontrollerで読み出してしまいたいかな、と思いました。いかがでしょうか?

ham-cap replied:
viewでSQLを発行したくない具体的な理由はどんなものがありますか？👀
DBに外部から干渉される隙になるとかでしょうか？
あとはページをブラウザに表示するのに時間がかかってしまうとかですか？🤔

--追記--
よく考えたら`@user.id`の部分に任意の値を渡したらそのユーザーの情報にアクセスできてしまいそうですね。
セキュリティ的に問題という認識で大丈夫でしょうか？👀

maedana replied:
viewに書く合理的な理由があればいいと思いますけど、そうでないならcontrollerに書いたほうが見通しがいいのではくらいです。今回の例では
```
current_user.active_relationships.find_by(followed_id: @user.id)
```
さえあればいいと思うので、controllerで設定してあげると見通しよくなるんじゃないかなと思ったんですよね。好みもありそうな部分ですけど。尚セキュリティ上の懸念はしてません。他者の個人情報が見れてしまうとかなら問題ですけど、別にそうではないですし。

maedana replied:
https://github.com/ham-cap/fjord-books_app/pull/5#discussion_r973823584 も合わせてご覧ください。

maedana commented on app/views/users/followers.html.erb:
```diff
@@ -0,0 +1,27 @@
+      </tr>
+    <% end %>
+  </tbody>
+</table>
```
N+1問題が起きてそうです。発生しないようにしましょう

maedana replied:
こちらいかがでしょうか?

ham-cap replied:
controller側でpreloadを使用して対応したつもりだったのですが、これでは不十分でしょうか？👀
↓こちらのコミットで修正している箇所です。
https://github.com/ham-cap/fjord-books_app/pull/5/commits/5e21ddbab24277a2baaec17876b7f6b003c93037

maedana replied:
失礼しました。見落としておりました。

なおActiveStorageで用意されているメソッドを利用するという手段もあります。以下もご覧頂くといいかもです。
https://api.rubyonrails.org/v7.0/classes/ActiveStorage/Attachment.html

maedana commented on app/views/devise/shared/_error_messages.html.erb:
```diff
@@ -3,8 +3,7 @@
                  count: resource.errors.count,
-                 resource: resource.class.model_name.human.downcase)
-       %>
+                 resource: resource.class.model_name.human.downcase) %>
```
変える必要あったんでしょうか? linterに引っかかったとかですか?

ham-cap replied:
ここはおっしゃるとおりlinterで引っかかりました。

maedana commented on app/models/user.rb:
```diff
@@ -5,4 +5,30 @@ class User < ApplicationRecord
+  end
+
+  def unfollow(other_user)
+    active_relationships.find_by(followed_id: other_user.id).destroy
```
find_byだと存在しないときにnilが返りますね。何が言いたいかというと複数タブ開いて既にフォロー解除済みで再度別タブでフォロー解除するとdestroyで500エラーになりそうです。その場合でも500エラーにならないようにしたいですね。(404エラーならOK)

maedana replied:
こちらいかがでしょう?

ham-cap replied:
こちら抜けていました。申し訳ありません。
対応します。

maedana commented on app/models/user.rb:
```diff
@@ -5,4 +5,30 @@ class User < ApplicationRecord
+    active_relationships.find_by(followed_id: other_user.id).destroy
+  end
+
+  def following?(other_user)
```
このメソッド使ってないような? そして実装的にこれではエラーになるような気が。

maedana commented on app/controllers/users_controller.rb:
```diff
@@ -8,4 +8,12 @@ def index
   end
+
+  def followings
+    @followings = User.find_by(id: params[:id]).followings
```
対象ユーザが存在しないときにfollowingsの呼び出しがエラーで500エラーになりそうです。404エラーとなるようにしたいですね。

maedana commented on app/controllers/users_controller.rb:
```diff
@@ -8,4 +8,12 @@ def index
+  end
+
+  def followers
+    @followers = User.find_by(id: params[:id]).followers
```
同上です。

ham-cap committed:
follwingsとfollowersメソッドで対象ユーザーが存在しない場合に404エラーが返るようにした

・対象となるユーザーの情報はfind_byで取得していたが、ユーザーが存在しない場合nilが返ってしまい、続くfollowingsとfollowersの呼び出しに失敗して500エラーになってしまうため、findを使用することで404エラーが返るようにした。

ham-cap committed:
開発環境でもブラウザ上で本番環境と同様のエラー画面が出力されるようにした

ham-cap committed:
フォロー解除の際に該当するactive_relationshipが存在しない場合に404エラーを返すようにした

・既にフォロー解除済の相手に対して、事前に複数タブで詳細ページにアクセスしておく等の方法で再度フォロー解除の操作を行った場合、destroyの呼び出しに失敗し500エラーになってしまうため、404エラーが返るようにした。

ham-cap committed:
不要なメソッドを削除

ham-cap committed:
フォロワー一覧画面を表示する際のN+1問題に対応

・ユーザーのアバター画像を読み込む際にSQLが都度発行されていたためコントローラー側でpreloadを使用して対応した。

ham-cap committed:
follow関連のルーティングからscopeを外した・フォロー/アンフォローの際にcontrollerでSQLを発行するようにした・フォローボタンの表示切り替えをif文に統一した

・follow関連のルーティングにscopeをつけることでidの重複等が見られ、わかりにくくなっていたため外した
・フォロー及びフォロー解除の際にviewでSQLの発行を行っていたが、できる限りコントローラー側で発行するように変更した
・current_userとアクセスしている詳細ページのユーザーが同じかどうかによって表示内容を切り替える際の処理をunlessとifで書いていたが、unlessを使う必要性がなかったためifに統一した

ham-cap committed:
followsテーブルの外部キーの設定方法をreferencesに変更しfollower_id及びfollowed_idカラムにnot null制約を追加

・followsテーブルではfollower_id、followed_idいずれも外部キーのため、referencesを使って設定した。
・follower_idとfollowed_idは必須のためnot null制約を追加した。

ham-cap committed:
使用していないファイルを削除

maedana commented on db/schema.rb:
```diff
@@ -50,12 +50,11 @@
     t.datetime "created_at", precision: 6, null: false
     t.datetime "updated_at", precision: 6, null: false
     t.index ["followed_id"], name: "index_follows_on_followed_id"
-    t.index ["follower_id", "followed_id"], name: "index_follows_on_follower_id_and_followed_id", unique: true
```
複合ユニークindex無くなってしまってますが、これは必要なので残さないとだめですね

maedana commented on config/environments/development.rb:
```diff
@@ -12,7 +12,7 @@
 
   # Show full error reports.
-  config.consider_all_requests_local = true
+  config.consider_all_requests_local = false
```
なぜ変えたんでしょう?

ham-cap replied:
500や404エラーが出た際に本番環境のブラウザ上でどのように表示されるのか確認したかったため変更しました。
ここを変えて以降のデバッグはターミナル上のログで行っていました。
開発環境だとブラウザにエラー内容が表示されるようにしておかないと不都合等ありますか？👀

maedana replied:
なるほどー、理解しました。OKです。ただ、個人的にはここは変更しないですね。目的を達成するなら/404とか/500を直接開けばいいのかなって思いました。

> 開発環境だとブラウザにエラー内容が表示されるようにしておかないと不都合等ありますか？eyes

パッと思いつかないので好みでいいんじゃないでしょうか。

ham-cap replied:
> 目的を達成するなら/404とか/500を直接開けばいいのかなって思いました。

たしかに仰るとおりですね🤔
次回以降はそうしようかと思います。

maedana commented on app/views/users/show.html.erb:
```diff
@@ -33,29 +33,26 @@
 </p>
   <strong><%= t('views.follows.followings') %>:</strong>
-  <%= link_to  @user.followings.count, followings_user_path %>
+  <%= link_to @the_number_of_followings , followings_user_path %>
```
前回指摘したことと矛盾するんですが、これくらいは普段controllerのインスタンス化してないですね。改めて考えてみたんですが、`ActiveRecordのインスタンスになるもの`はcontrollerで作るように意識していますね。フォロー/アンフォローでは問題にならなそうですが、例えばvalidation errorになったときにこのviewをrenderすることを考えた場合、viewでfindするということはエラー情報が維持出来なくて困ったりしますし。

ham-cap replied:
ということは、少なくとも今回は変更前の`@user.followings.count`のままにしておくほうが良いという認識であっていますか？👀
また、ActiveRecordのインスタンスになるものというのは、例えば今回でいうとusers/show.html.erbでユーザーをフォローする際に`form_with`に渡している`@follow`のようなイメージで大丈夫でしょうか？（`@follow`は`follower_id`が空の状態のactive_relationshipのインスタンスです。）

maedana replied:
> ということは、少なくとも今回は変更前の@user.followings.countのままにしておくほうが良いという認識であっていますか？

そうですねー

> また、ActiveRecordのインスタンスになるものというのは、例えば今回でいうとusers/show.html.erbでユーザーをフォローする際にform_withに渡している@followのようなイメージで大丈夫でしょうか？（@followはfollower_idが空の状態のactive_relationshipのインスタンスです。）

今回でいうと以下3つですかね
```
@user, @following, @follow
```

maedana commented on app/views/users/show.html.erb:
```diff
@@ -33,29 +33,26 @@
-    <%= form_with(model: current_user.active_relationships.find_by(followed_id: @user.id),
-                                                                          html: { method: :delete },
-                                                                         local: true) do |f| %>
+    <%= form_with(model: @following, method: :delete, local: true) do |f| %>
```
```
  <% if current_user.followings.exists?(id: @user.id) %>
    <%= form_with(model: @following, method: :delete, local: true) do |f| %>
```
が、2行なくても`@following,`有無だけで判断していいと思うのですが、いかがでしょう?

ham-cap replied:
たしかにそうですね。
ちょっと書き方を変えてみます。

ham-cap committed:
Userモデル内のunfollowメソッドにおいてactive_relationshipsが見つからない場合に404エラーを返すようにした

・active_relationshipsから対象ユーザーを取得する際にfind_byだと対象が見つからない場合にnilが返ってしまうため、find_by!を使用して例外を発生させるようにした。

ham-cap committed:
followsテーブルのfollower_idとfollowed_idに複合ユニーク制約を設定した

・followsテーブルのレコードにおいてはfollower_idとfollowed_idは2つ揃っていないと意味がなく、各カラムのユニーク制約に加えて複合ユニーク制約も設定していたが誤って削除してしまっていたため再度設定した。

ham-cap committed:
フォロー数とフォロワー数をview側でカウントして表示するようにした・フォロー/フォロー解除ボタンのいずれかを表示させるif文を修正した

・フォロー数とフォロワー数を表示するlink_toにはcontroller側でカウントした数値をインスタンス化して渡していたが、ActiveRecordのインスタンスにならないもののためview側でカウントすることにした
・current_userと@userが同じかどうかを判定するif文の中で更にcurrent_userのフォローを呼び出して@userが含まれているかを判定していたが、@followingの存在確認を最初のif文に含めることでif文のネストをなくした

maedana commented on app/views/users/show.html.erb:
```diff
@@ -33,26 +33,24 @@
+    <div><%= hidden_field_tag :followed_id, @user.id %></div>
+    <%= f.submit t('views.follows.follow_button') %>
+  <% end %>
+<% elsif current_user == @user %>
```
current_userの判定が複数回出てくるのとなるべく否定の`!`使わないほうが読みやすいかなって感じます。僕だったら
```ruby
if current_user == @user
  ...
else
  if @following
    ...
  else
    ...
  end
end
```
と分岐させるかなーと思いました。

ham-cap replied:
なるほど👀
たしかにこのほうがシンプルで読みやすいですね！
頭が固くなっていました。
